File: image_processing/general_implementations/halftoning_rgb_v1.m
function halftoning_rgb_example(image_path)
    % Read an image
    img = imread('input/im1');
    
    % Apply Floyd-Steinberg error diffusion
    img_error_diffusion = floyd_steinberg_error_diffusion(img);
    figure, imshow(img_error_diffusion), title('Error Diffusion (Floyd-Steinberg)');
    
    % Apply Ordered Dithering
    img_ordered_dithering = ordered_dithering(img);
    figure, imshow(img_ordered_dithering), title('Ordered Dithering (Bayer Matrix)');
end

%% Floyd-Steinberg Error Diffusion Function
function output = floyd_steinberg_error_diffusion(img)
    % Convert image to double for calculation
    img = double(img);
    [rows, cols, channels] = size(img);
    
    % Floyd-Steinberg diffusion matrix coefficients
    diffusion_matrix = [0 0 7; 3 5 1] / 16;
    
    % Loop over each pixel in the image
    for y = 1:rows
        for x = 1:cols
            for c = 1:channels
                % Get original pixel value
                old_pixel = img(y, x, c);
                
                % Threshold to either 0 or 255
                new_pixel = round(old_pixel / 255) * 255;
                img(y, x, c) = new_pixel;
                
                % Compute error
                error = old_pixel - new_pixel;
                
                % Diffuse error to neighboring pixels
                for dy = 0:1
                    for dx = -1:1
                        if (x + dx > 0) && (x + dx <= cols) && (y + dy <= rows)
                            img(y + dy, x + dx, c) = img(y + dy, x + dx, c) + error * diffusion_matrix(dy + 1, dx + 2);
                        end
                    end
                end
            end
        end
    end
    
    % Convert the image back to uint8 and return
    output = uint8(img);
end

%% Ordered Dithering Function
function output = ordered_dithering(img)
    % Convert image to double for calculation
    img = double(img);
    
    % Define a 4x4 Bayer matrix
    bayer_matrix = [0 8 2 10; 12 4 14 6; 3 11 1 9; 15 7 13 5] * (255 / 16);
    [rows, cols, channels] = size(img);
    [bm_rows, bm_cols] = size(bayer_matrix);
    
    % Loop over each pixel and apply the Bayer matrix
    for y = 1:rows
        for x = 1:cols
            for c = 1:channels
                % Get the threshold from the Bayer matrix
                threshold = bayer_matrix(mod(y-1, bm_rows) + 1, mod(x-1, bm_cols) + 1);
                
                % Apply thresholding
                if img(y, x, c) > threshold
                    img(y, x, c) = 255;
                else
                    img(y, x, c) = 0;
                end
            end
        end
    end
    
    % Convert the image back to uint8 and return
    output = uint8(img);
end

File: image_processing/halftoning_rgb/halftoning_rgb_v1.py
import numpy as np
from PIL import Image

"""
Error Diffusion algorithm:
a. For each pixel in the image:
   - Add the accumulated error to the current pixel value.
   - Compare the result to a threshold to determine if it should be white (255) or black (0).
   - Calculate the quantization error (difference between the old and new pixel values).
   - Distribute the error to the neighboring pixels that haven't been processed yet.
- We use the Floyd-Steinberg distribution pattern: 7/16 to the right, 3/16 to the bottom-left, 5/16 to the bottom, and 1/16 to the bottom-right.
- The error is treated as a vector (RGB), so each color channel is processed independently.


Dithering algorithm:
a. For each pixel in the image:
   - Calculate the corresponding threshold value from the Bayer matrix.
   - Compare the pixel value to the threshold to determine if it should be white (255) or black (0).
- We use a 4x4 Bayer matrix, which is tiled across the entire image.
- The matrix values are normalized to the range [0, 1] and then scaled to [0, 255] for comparison with pixel values.
"""

def error_diffusion(image, threshold=128):
    """
    Apply error diffusion halftoning to an RGB image.
    
    Args:
    image (numpy.ndarray): Input RGB image (height, width, 3)
    threshold (int): Threshold value for binarization (0-255)
    
    Returns:
    numpy.ndarray: Halftoned image
    """
    height, width, _ = image.shape
    output = np.zeros_like(image)
    error = np.zeros_like(image, dtype=float)
    
    for y in range(height):
        for x in range(width):
            old_pixel = image[y, x].astype(float) + error[y, x]
            new_pixel = np.where(old_pixel > threshold, 255, 0)
            output[y, x] = new_pixel
            
            quant_error = old_pixel - new_pixel
            
            if x + 1 < width:
                error[y, x + 1] += quant_error * 7/16
            if x - 1 >= 0 and y + 1 < height:
                error[y + 1, x - 1] += quant_error * 3/16
            if y + 1 < height:
                error[y + 1, x] += quant_error * 5/16
            if x + 1 < width and y + 1 < height:
                error[y + 1, x + 1] += quant_error * 1/16
    
    return output.astype(np.uint8)

def ordered_dithering(image):
    """
    Apply ordered dithering halftoning to an RGB image.
    
    Args:
    image (numpy.ndarray): Input RGB image (height, width, 3)
    
    Returns:
    numpy.ndarray: Halftoned image
    """
    height, width, _ = image.shape
    output = np.zeros_like(image)
    
    # Bayer matrix 4x4
    bayer_matrix = np.array([
        [ 0, 8, 2, 10],
        [12, 4, 14, 6],
        [ 3, 11, 1, 9],
        [15, 7, 13, 5]
    ]) / 16
    
    for y in range(height):
        for x in range(width):
            threshold = bayer_matrix[y % 4, x % 4] * 255
            output[y, x] = np.where(image[y, x] > threshold, 255, 0)
    
    return output.astype(np.uint8)

def main():
    input_image = np.array(Image.open("input/image1.png"))
    
    # Apply error diffusion
    error_diffusion_result = error_diffusion(input_image)
    Image.fromarray(error_diffusion_result).save("error_diffusion_result.png")
    
    # Apply dithering
    dithering_result = ordered_dithering(input_image)
    Image.fromarray(dithering_result).save("output/dithering_result.png")


if __name__ == "__main__":
    main()

